tests/worker/test_workflow.py

```diff
@@ -7622,3 +7628,201 @@ async def test_workflow_missing_local_activity_no_activities(client: Client):
+            assert result[0] == temporalio.activity.ActivityCancellationDetails(
+                paused=True
+            )
+            assert result[1] == temporalio.activity.ActivityCancellationDetails(
+                paused=True
+            )
+
+
+@workflow.defn
+class ActivityHeartbeatPauseUnpauseWorkflow:
+    @workflow.run
+    async def run(
+        self, activity_id: str
+    ) -> list[Optional[temporalio.activity.ActivityCancellationDetails]]:
+        results = []
+        results.append(
+            await workflow.execute_activity(
+                sync_heartbeat_activity,
+                False,
+                activity_id=activity_id,
+                start_to_close_timeout=timedelta(seconds=10),
+                heartbeat_timeout=timedelta(seconds=1),
+                retry_policy=RetryPolicy(maximum_attempts=2),
+            )
+        )
+        results.append(
+            await workflow.execute_activity(
+                heartbeat_activity,
+                False,
+                activity_id=f"{activity_id}-2",
+                start_to_close_timeout=timedelta(seconds=10),
+                heartbeat_timeout=timedelta(seconds=1),
+                retry_policy=RetryPolicy(maximum_attempts=2),
+            )
+        )
+        return results
+
+
+async def test_activity_pause_unpause(client: Client, env: WorkflowEnvironment):
+    if env.supports_time_skipping:
+        pytest.skip("Time-skipping server does not support pause API yet")
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        async with Worker(
+            client,
+            task_queue=str(uuid.uuid4()),
+            workflows=[ActivityHeartbeatPauseUnpauseWorkflow],
+            activities=[heartbeat_activity, sync_heartbeat_activity],
+            activity_executor=executor,
+            max_heartbeat_throttle_interval=timedelta(milliseconds=300),
+            default_heartbeat_throttle_interval=timedelta(milliseconds=300),
+        ) as worker:
+            test_activity_id = f"heartbeat-activity-{uuid.uuid4()}"
+
+            handle = await client.start_workflow(
+                ActivityHeartbeatPauseUnpauseWorkflow.run,
+                test_activity_id,
+                id=f"test-activity-pause-unpause-{uuid.uuid4()}",
+                task_queue=worker.task_queue,
+            )
+
+            # Wait for sync activity
+            activity_info_1 = await assert_pending_activity_exists_eventually(
+                handle, test_activity_id
+            )
+            # Assert not paused
+            assert not activity_info_1.paused
+            # Pause activity then assert it is paused
+            await pause_and_assert(client, handle, activity_info_1.activity_id)
+
+            # Wait for next heartbeat to propagate the cancellation. Unpausing before the heartbeat
+            # will show activity as unpaused to core. Consequently, it will *not* issue an activity cancel.
+            time.sleep(0.3)
```

Comment:
Is there a way to get rid of potentially flaky sleeps and wait for situations that are needed? For instance in https://github.com/temporalio/sdk-dotnet/pull/482 I used the presence of the final heartbeat details to confirm activity completed (kinda).

Reply:
replaced with a helper that waits for the next heartbeat to elapse